src/Compilers/CSharp/Portable/Binder/Binder_Await.cs

```diff
@@ -307,51 +306,120 @@ private bool GetAwaitableExpressionInfo(
-                        new ConstraintsHelper.CheckConstraintsArgs(this.Compilation, this.Conversions, includeNullability: false, expression.Syntax.Location, diagnostics));
+                    if (member is not MethodSymbol method
+                        || method.Arity != exprType.Arity
+                        || method.ParameterCount > 1)
+                    {
+                        continue;
+                    }
+
+                    if (method.Arity == 0)
+                    {
+                        if (method.ReturnsVoid && isValidConversion(exprType, method, node, diagnostics, this))
+                        {
+                            if (runtimeAwaitHelper is null)
+                            {
+                                runtimeAwaitHelper = method;
+                                continue;
+                            }
+                            else
+                            {
+                                runtimeAwaitHelper = null;
+                                return false;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        var unsubstitutedReturnType = method.ReturnType;
+                        if ((object)unsubstitutedReturnType != method.TypeArgumentsWithAnnotations[0].Type)
+                        {
+                            continue;
+                        }
+
+                        var substitutedMethod = method.Construct(exprType.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics);
+                        var tempDiagnostics = BindingDiagnosticBag.GetInstance(diagnostics);
+                        if (!ConstraintsHelper.CheckConstraints(
+                            substitutedMethod,
+                            new ConstraintsHelper.CheckConstraintsArgs(this.Compilation, this.Conversions, includeNullability: false, expression.Syntax.Location, tempDiagnostics)))
+                        {
+                            tempDiagnostics.Free();
+                            continue;
+                        }
+
+                        if (!isValidConversion(exprType, substitutedMethod, node, diagnostics, this))
+                        {
+                            tempDiagnostics.Free();
+                            continue;
+                        }
+
+                        diagnostics.AddRangeAndFree(tempDiagnostics);
+
+                        if (runtimeAwaitHelper is null)
+                        {
+                            runtimeAwaitHelper = substitutedMethod;
+                        }
+                        else
+                        {
+                            runtimeAwaitHelper = null;
+                            return false;
+                        }
+                    }
+
+                    static bool isValidConversion(TypeSymbol exprOriginalType, MethodSymbol method, SyntaxNode node, BindingDiagnosticBag diagnostics, Binder @this)
```

Comment:
This local function is in the middle of a local function body (ie. there's a return below). I believe that's something we avoid or maybe even have a guideline against